Move contracted defns to a submodule.

samth opened the pull request:
I'll squash this and write a real commit message once it works.
Which it doesn't yet.

- [x] pass unit tests
- [x] pass integration tests
- [x] build math and plot
- [x] write a comment explaining this whole thing
- [x] check for excess dependencies
- [x] write a decent commit message

samth commented:
This is now in very close to final form. The remaining problem is that static struct info doesn't provide the correct version of the struct identifiers, but everything else appears to build.

@takikawa If you want to start building on this, it's probably reasonable now.

andmkent commented:
Looks fine to me. (caveat: I've done no hacking in this part of TR yet)

takikawa commented on typed-racket-lib/typed-racket/private/type-contract.rkt:
```diff
@@ -157,15 +160,21 @@
+        #:when (contract-def/provide-property form)
+        (generate-contract-def/provide form ctc-cache sc-cache)]
+       [(module* name #f forms ...)
+        #`(module* name #f 
```
Would preserving syntax locations via `syntax/loc` matter for either of these cases?

samth replied:
I don't think it can matter, but I'll do it.

takikawa commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -26,7 +26,7 @@
+;; generate-prov : dict[id -> def-binding] dict[id -> list[id]] id id
 ;;                 -> (values listof[syntax] listof[listof[list[id id]]])
 ;; defs: defines in this module
 ;; provs: provides in this module
```
Can you add a description of the `mk-redirect-id` parameter too?

samth replied:
Done.

takikawa commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -100,11 +101,15 @@
                    [export-id new-id]
-                   [protected-id (freshen-id #'id)])
+                   [protected-id (freshen-id #'id)]
+                   [local-untyped-id (freshen-id #'id)])
```
Is this identifier unused?

takikawa commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -112,29 +117,31 @@
-            (lambda (stx)
-              (tc-error/stx stx "Macro ~a from typed module used in untyped code" 'untyped-id))))
+                   [untyped-id (freshen-id #'id)]
+                   [local-untyped-id (freshen-id #'id)])
```
Ditto, unused?

takikawa commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -37,17 +37,18 @@
-  ;; First return value is a syntax object of definitions
-  ;; Second is the id to export
-  ;; Third is a list of two element lists representing type aliases
+  ;; quad/c in the signatures corresponds to four values:
```
Just want to make sure I understand this change. So we need to split up the definition vs. just the id to export so that the definitions can be relocated into a submodule right?

samth replied:
Yes, that's right. I've improved the comment, hopefully.

takikawa commented on typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:
```diff
@@ -6,15 +6,15 @@
-         (typecheck provide-handling def-binding tc-structs
-                    typechecker internal-forms)
+         "provide-handling.rkt" "def-binding.rkt" "tc-structs.rkt"
+         "typechecker.rkt" "internal-forms.rkt"
```
Is there any reason to make these relative paths here? (and if so, maybe a separate commit?)

samth replied:
Relative paths work better with check syntax in drracket.

takikawa commented on typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:
```diff
@@ -6,15 +6,15 @@
+         "typechecker.rkt" "internal-forms.rkt"
          syntax/location
-         racket/format
+         racket/format typed-racket/utils/redirect-contract
```
Can this go in the `(utils ...)` require above?

takikawa commented on typed-racket-lib/typed-racket/utils/redirect-contract.rkt:
```diff
@@ -0,0 +1,25 @@
+(require syntax/modcollapse (for-template racket/base))
+(provide make-make-redirect-to-contract)
+
+(define (((make-make-redirect-to-contract varref) id) stx)
```
I think this module/function could use purpose statements. Especially since `make-make-redirect-to-contract` is a scary function name.

samth commented:
I've addressed most of the comments, but I still need to write a big doc comment for @stamourv about what the submodule is doing. More feedback before I get that done is also welcome.

samth commented:
@stamourv let me know what you think about that explanation

samth commented:
Turns out that this won't entirely remove the dependencies it should, because `prims.rkt` depends on too much stuff. For example, the implementation of `require/typed` has a direct dependency on `racket/contract/base`, which we should remove. I'll open a separate issue for reducing prims dependencies generally.

samth commented:
This is now done and ready to merge from my perspective.

shekari commented on typed-racket-lib/typed-racket/private/type-contract.rkt:
```diff
@@ -157,15 +164,25 @@
-(define (change-provide-fixups forms)
-  (define ctc-cache (make-hash))
-  (define sc-cache (make-hash))
+;; These are probably all in a specific place, which could avoid the big traversal
```
This should be marked as a TODO.

shekari commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -62,35 +66,38 @@
                  ;; if it wasn't there, put it in, and skip this case
                  (λ () (dict-set! mapping internal-id new-id) #f))
-       => (λ (mapped-id) (values #'(begin) mapped-id null))]
+       => (λ (mapped-id) (values #'(begin) #'(begin) mapped-id null))]
```
This seem like it is the same as line 80. Maybe add a function: `(define (mk-ignored-quad id) (values #'(begin) #'begin) id null))`. Not sure if that would be cleaner though.

shekari commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -62,35 +66,38 @@
                   (lambda (e) (if (identifier? e)
                                   (mk e)
-                                  (values #'(begin) e null)))
+                                  (values #'(begin) #'(begin) e null)))
```
Could reuse `mk-ignored-quad` here as well if it supported `#f`.

shekari commented on typed-racket-lib/typed-racket/typecheck/provide-handling.rkt:
```diff
@@ -112,47 +132,58 @@
-        #`(rename-out [#,id #,external-id])))
-    (when (pair? external-ids) (set-box! include-extra-requires? #t))
-    (values #`(begin #,defs (provide #,@provide-forms))
+        (if (free-identifier=? id external-id)
```
Why do we need to check here? Is this just an optimization?

samth replied:
It just simplifies the expansion.

shekari commented on typed-racket-lib/typed-racket/utils/redirect-contract.rkt:
```diff
@@ -0,0 +1,45 @@
+(require syntax/modcollapse (for-template racket/base))
+(provide make-make-redirect-to-contract)
+
+;; This produces an identifier that expands to a local-require.  It's
```
This sentence confused me, since nothing here returns an identifier. It seems that make-make-redirect-to-contract is a curried function that takes `varref` and `id` and returns a syntax transformer that expands to a local require.

shekari commented on typed-racket-lib/typed-racket/utils/redirect-contract.rkt:
```diff
@@ -0,0 +1,45 @@
+
+;; stx: the syntax object that's the argument to the macro (that is,
+;;      the stx object that's the reference to the typed identifier in
+;;      a untyped module). The funny eta-expansion with `redirect` is
```
I don't see why you need the eta expansion, just remove the `stx` argument and return `redirect`. This also makes it clearer that this is returning a syntax transformer.

shekari commented on typed-racket-lib/typed-racket/utils/redirect-contract.rkt:
```diff
@@ -0,0 +1,45 @@
+                             varref)))]
+                    [i (datum->syntax id (syntax-e id) stx stx)])
+        #`(let ()
+            (local-require (only-in mp [#,(datum->syntax #'mp (syntax-e #'i)) i]))
```
I'm confused by the context argument to datum->syntax here. Can you explain whats going on? I thought module exports were symbols not identifiers, and so this should be unnecessary.

samth replied:
Module exports are symbols, but `require` and `local-require` bind based on identifiers. This makes sure that the name binds the reference on the next line.

shekari commented on typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:
```diff
@@ -350,63 +362,130 @@
+              (dict-update h #'i (lambda (tail) (cons #'i tail)) '())]
+             [((~datum rename) in out)
+              (dict-update h #'in (lambda (tail) (cons #'out tail)) '())]
+             [(name:unknown-provide-form . _)
```
Should these be unsupported provide form instead of unknown?

samth replied:
You mean rename the syntax class? There isn't anything currently called `unsupported-provide-form`.

shekari commented:
LGTM.

stamourv commented:
Sorry for the late reply, was traveling.
Comments look good. We need more like that.
